### PR TITLE
Initial BLS Beacon

### PR DESCRIPTION
**Pull request by ejfitzgerald**

### CODE REVIEW

File: libs/crypto/examples/dkg/main.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+  // Generating secret and collecting secrets.
+  std::vector<VerificationVector> verification_vectors;
+  for (uint64_t i = 0; i < members.size(); ++i)
+  {
+    auto contrib = bls::dkg::GenerateContribution(participants, threshold);
+    // Note that the verfication vector can be posted publicly.
+    verification_vectors.push_back(contrib.verification);
+
+    for (uint64_t i = 0; i < contrib.contributions.size(); ++i)
```
Comment (troels-im) [inline libs/crypto/examples/dkg/main.cpp:85]: inner loop should be over `j`

File: libs/crypto/include/crypto/bls.hpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+#include <atomic>
+#include <sstream>
+#include <utility>
+
+namespace fetch {
+namespace crypto {
+
+namespace details {
+// TODO: This entire file should be updated using the `bls_base.hpp` instead.
```
Comment (troels-im) [inline libs/crypto/include/crypto/bls.hpp]: Needs an issue and issue number

File: libs/crypto/include/crypto/bls_base.hpp [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  E_FR_SIZE              = E_MCLBN_FP_UNIT_SIZE * 8,
+  E_ID_SIZE              = E_FR_SIZE,
+  E_G1_SIZE              = E_FR_SIZE * 3,
+  E_G2_SIZE              = E_FR_SIZE * 3 * 2
+};
+
+inline void Init()
+{
+
+  // TODO: add atomic lock.
```
Comment (troels-im) [inline libs/crypto/include/crypto/bls_base.hpp]: create issue, or copy blueprint from `bls.hpp`

File: libs/dkg/include/dkg/dkg_service.hpp [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+ *                       └───────────────────────┘            │
+ *                                   │                                     │
+ *                                   │                        │
+ *                                   ▼                                     │
+ *                       ┌───────────────────────┐            │
+ *                       │                       │                         │
+ *                       │       Complete        │─ ─ ─ ─ ─ ─ ┴ ─ ─ ─ ─ ─ ─
+ *                       │                       │
+ *                       └───────────────────────┘
+ */
```
Comment (troels-im) [inline libs/dkg/include/dkg/dkg_service.hpp:104]: Awesome!

File: libs/ledger/include/ledger/consensus/stake_manager.hpp [hunk 5]
```diff
@@ -70,17 +75,20 @@ class StakeManager final : public StakeManagerInterface
...
+  using EntropyCache     = std::map<BlockIndex, uint64_t>;
 
   StakeSnapshotPtr LookupStakeSnapshot(BlockIndex block);
   void             ResetInternal(StakeSnapshotPtr &&snapshot, std::size_t committee_size);
+  bool             LookupEntropy(Block const &block, uint64_t &entropy);
 
   // Config & Components
   std::size_t                committee_size_{0};       ///< The "static" size of the committee
-  EntropyGeneratorInterface &entropy_;                 ///< The reference to entropy module
+  EntropyGeneratorInterface *entropy_;                 ///< The reference to entropy module
```
Comment (troels-im) [inline libs/ledger/include/ledger/consensus/stake_manager.hpp]: Default initialisation to `nullptr`?

File: apps/constellation/constellation.hpp [hunk 6]
```diff
@@ -178,6 +185,7 @@ class Constellation : public ledger::BlockSinkInterface
 
   /// @name Staking
   /// @{
+  DkgServicePtr   dkg_;      ///> The DKG system
```
Comment (troels-im) [inline apps/constellation/constellation.hpp]: `//<`?

File: libs/crypto/examples/dkg/main.cpp (hunk 1 above)
Comment (ejfitzgerald) [reply libs/crypto/examples/dkg/main.cpp:85]: #1286

**Pull request merged**